src/PluginStartupRegistration.cpp

```diff
@@ -0,0 +1,81 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file PluginStartupRegistration.cpp
+
+  @author Vitaly Sverchinsky
+
+**********************************************************************/
+
+#include "PluginStartupRegistration.h"
+
+#include <thread>
+
+#include <wx/log.h>
+#include <wx/app.h>
+
+#include "PluginManager.h"
+#include "PluginDescriptor.h"
+#include "widgets/ProgressDialog.h"
+
+
+PluginStartupRegistration::PluginStartupRegistration(std::vector<std::pair<wxString, wxString>> plugins) : mPlugins(std::move(plugins))
+{
+   mValidator = std::make_unique<AsyncPluginValidator>(*this);
+}
+
+void PluginStartupRegistration::OnInternalError(const wxString& error)
+{
+   wxLogError(error);
+   Stop();
+}
+
+void PluginStartupRegistration::OnPluginFound(const PluginDescriptor& desc)
+{
+   PluginManager::Get().RegisterPlugin(PluginDescriptor { desc });
+   if(!desc.IsValid())
+      mFailedPlugins.push_back(desc.GetPath());
+}
+
+void PluginStartupRegistration::OnValidationFinished()
+{
+   ++mNumChecked;
+   ProcessNext();
+}
+
+const std::vector<wxString>& PluginStartupRegistration::GetFailedPlugins() const noexcept
+{
+   return mFailedPlugins;
+}
+
+void PluginStartupRegistration::Run()
+{
+   mDialog = BasicUI::MakeProgress(XO("Searching for plugins"), XO(""));
+   ProcessNext();
+   while(mValidator)
+   {
+      const auto message = TranslatableString { mPlugins[mNumChecked].second, { } };
+      //Update UI
+      if(mDialog->Poll(mNumChecked, mPlugins.size(), message) != BasicUI::ProgressResult::Success)
+         Stop();
+      //Process pending events
+      wxTheApp->Yield();
```

Comment:
`Poll` already did a `Yield` too

Reply:
It will cost some time because from what I see it doesn't yield each `Poll`. In my case the difference was 2-4 seconds.